services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,306 @@
+const {ApiError, Provider} = require('./provider');
+const aws = require('aws-sdk');
+const taskcluster = require('taskcluster-client');
+const crypto = require('crypto');
+const fs = require('fs');
+const path = require('path');
+const assert = require('assert');
+const {AWS_API_VERSION} = require('../constants');
+
+class AwsProvider extends Provider {
+  constructor({
+    providerId,
+    monitor,
+    rootUrl,
+    Worker,
+    WorkerPool,
+    WorkerPoolError,
+    estimator,
+    validator,
+    notify,
+    providerConfig,
+  }) {
+    super({
+      providerId,
+      monitor,
+      rootUrl,
+      Worker,
+      WorkerPool,
+      WorkerPoolError,
+      estimator,
+      validator,
+      notify,
+    });
+    this.configSchema = 'config-aws';
+    this.ec2iid_RSA_key = fs.readFileSync(path.resolve(__dirname, 'aws-keys/RSA-key-forSignature')).toString();
+    this.providerConfig = providerConfig;
+  }
+
+  /*
+   This method is used to setup permissions for the EC2 instances
+   */
+  async setup() {
+  }
+
+  async provision({workerPool}) {
+    const {workerPoolId} = workerPool;
+
+    if (!workerPool.providerData[this.providerId] || workerPool.providerData[this.providerId].running === undefined) {
+      await workerPool.modify(wt => {
+        wt.providerData[this.providerId] = wt.providerData[this.providerId] || {};
+        wt.providerData[this.providerId].running = wt.providerData[this.providerId].running || 0;
+      });
+    }
+
+    const config = this.chooseConfig({possibleConfigs: workerPool.config.launchConfigs});
+
+    aws.config.update({region: config.region});
+    aws.config.logger = console;
+    const ec2 = new aws.EC2({
+      apiVersion: AWS_API_VERSION,
+      credentials: this.providerConfig.credentials,
+    });
+
+    const toSpawn = await this.estimator.simple({
+      workerPoolId,
+      minCapacity: config.launchConfig.MinCount,
+      maxCapacity: config.launchConfig.MaxCount,
+      capacityPerInstance: config.capacityPerInstance,
+      running: workerPool.providerData[this.providerId].running,
+    });
```

Comment:
If `toSpawn` is zero should this exit here?

Reply:
according to my improved understanding - yes!